### DISCUSSION TITLE
Adding support for custom board/processor

### DISCUSSION THREAD

**Discussion by usmnzain** [eclipse-embed-cdt/eclipse-plugins#529] [General Discussions]
I am trying to add support for the custom SoC that I have built. The eclipse-embed project supports some hifive boards that you can selected when creating a new C/C++ project. I was wondering how can I see my board in that list. I tried looking on relevant sites but can not figure out where to start from. 

I have previously used platform io to program my board (added support for my custom SoC) so I am not a complete newbie.

**Comment by ilg-ul** [comment]
Adding project templates is not a trivial task.

First you have to create the project as you want it to result, and be sure it compiles and runs properly. Then you have to create a separate project template plug-in, which includes the files plus the commands to create the desired configuration. 

If the template has different options (like the SiFive templates), projects get really complicated, you have to use a subtitution mechanism to customise the content of the files based on the configuration options.

I suggest you start with the first step, to create the project as you want it to be. Such a project can be later imported directly into Eclipse, and used as-is, you don't need a full template plug-in for this.

**Reply by usmnzain** [reply]
If i create the project would it save all the debugging configs and capabilities? 

If i want my board to appear here; inside the project type window so that when anyone creates a new project they have an option to select my board, what steps would be required, briefly.  
![](https://eclipse-embed-cdt.github.io/assets/images/2015/intro-new-f4-project.png)

**Reply by ilg-ul** [reply]
> would it save all the debugging configs and capabilities

It is up to you to save the debug launchers as files in the project or inside Eclipse; see the **Common** tab.

> If i want my board to appear here ... what steps would be required, briefly.

You have to write a plug-in like these:

- <https://github.com/eclipse-embed-cdt/eclipse-plugins/tree/master/plugins/org.eclipse.embedcdt.templates.sifive.ui>
- <https://github.com/eclipse-embed-cdt/eclipse-plugins/tree/master/plugins/org.eclipse.embedcdt.templates.stm.ui>

Then you have to publish it somewhere, and allow users to **Install New Software...** from there.

**Reply by usmnzain** [reply]
Thanks @ilg-ul for the pointers. I will start with your suggestion first by creating a project according to my requirements and hosting it on github.

**Reply by ilg-ul** [reply]
As for customisation options, if you decide for a template that generates projects with various configuration options, you have two choices: a simple variable substitution mechanism, available in CDT, with a syntax like `$(variable)`, and a powerful template engine (Liquid) that was used in the SiFive plug-in.

**Reply by usmnzain** [reply]
Hello! Initially I am attempting to just setup a basic project that includes the drivers for the peripherals and verifying if builds everything correctly. So far, the project builds properly with my linker, startup code and drivers and is able to debug my target (running in simulation). 

Next, I am trying to somehow flash the the elf to my SoC running on fpga (after converting to hex). After some searching, I see few examples of expressif esp-idf eclipse plugin that allows to flash their esp32 through eclipse. So, does eclipse provide such interface and if not, how should progress if i want an interface to flash my board.

**Reply by ilg-ul** [reply]
I have no experience with Espressif. For flashing/debugging my devices I use J-Link and occasionally OpenOCD, via the corresponding Eclipse plug-ins. However, if you have a command line tool for flashing your devices, you can invoke it as an external tool from Eclipse, you don't need a dedicated plug-in only for that.
